### PR TITLE
Rewards: Update Rewards verified sign by fetching publisher info on url change

### PR DESCRIPTION
**Pull request by danishjafri88**
Fixes#: brave/brave-rewards-ios#225

The rewards verified sign was not updated when browsing sites without page load, example YouTube.
## Summary of Changes
A fetch publisher activity is done in delegate to update url. This internally checks if its media type or normal and does the required fetch -> call observer. We already have code in place to update the UI in observer.
This pull request fixes issue #<number>

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

Open a verified YouTube page and verify the rewards verified sign is displayed
Visit another video from the same page that is not verified and check that the rewards verified sign is removed.

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/internal/issues/new?assignees=&labels=security%2C+security-reviews&template=security-review.md&title=iOS:%20Security+Review+for) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift]: This logic change worries me a bit. It used to be if the URL was nil for whatever reason, we'd be emptying the current url and do other stuff below, but now you guard the entire method based on existence of that URL. Can we maybe isolate this change-set to not alter other logic in this function?

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/BrowserViewController.swift
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift]: This logic is still different. I think this is what you're looking for:

```swift
if let url = tab.url {
  if url.isLocal {
    self.topToolbar.locationView.rewardsButton.isVerified = false
  } else {
    self.rewards?.ledger.fetchPublisherActivity(from: url, faviconURL: nil, publisherBlob: nil, tabId: UInt64(tab.rewardsId))
  }
}
```

Also:
1. Is it possible to get the faviconURL here? 
2. Can you double check that `twitch.tv` still shows publisher names in the panel when you switch pages? (As it uses publisher blob)

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/BrowserViewController.swift
Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController.swift]: For the case if its local url or the url is nil the isVerified should be set false. In the other case it should fetch. 
The `if url.isLocal {
    self.topToolbar.locationView.rewardsButton.isVerified = false
  }` is also my code from a previous pr.

In your recommended code the verified sign will still be there if the url is set to nil

Comment (kylehickinson) [reply Client/Frontend/Browser/BrowserViewController.swift]: > In your recommended code the verified sign will still be there if the url is set to nil

This is how the logic is _now_, before this change, based on how the code is written. Was the change on purpose you mean then? To make it go away if the URL is nil?

Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController.swift]: Yeah, I realised this case when working on this pr. It should be set to false if the url is nil too. Thus corrected it in this pr. 

For twitch.tv we definitely need the blob, looking at how to fix for it too. Thanks for lmk

**Comment by danishjafri88** [comment]
@kylehickinson I tested the new code, with YouTube, twitch, GitHub & reddit (could not verify for twitter as I could not find a verified publisher). 

Albeit there is a redundancy:
the `reportPageLoad` call is done from `func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {` too. After a very thorough testing I found that almost always the publisher can be found where I make call (URL change). If you think its ok maybe we can remove the reporting in `didFinishNavigation`

**Comment by kylehickinson** [comment]
Hmm, I don't know if we should be calling `reportPageLoaded` as that has other implications (classifying for ads for instance). @tmancey @NejcZdovc Any idea what we should do here? Mobile YouTube.com doesn't report navigations when you go from page to page because its using XHR loads and JS, so we're not getting reports when things "finish" from WKWebView.

**Comment by tmancey** [comment]
@kylehickinson @danishjafri88 Will wait for feedback after your call, thanks

**Comment by kylehickinson** [comment]
Going to KVO on URL change, check if the URL is a media pub when that URL changes, and if its mobile youtube/mobile twitch we will call ledger's OnUnload, then ledgers OnLoad, ads OnPageLoad. These special calls will only exist for these two media publisher websites

**Comment by danishjafri88** [comment]
@kylehickinson rewrote the logic with discussed changes. Tested with YouTube and twitch and its works good. Please test it out yourself too :)

File: Shared/Extensions/URLExtensions.swift
Comment (kylehickinson) [inline Shared/Extensions/URLExtensions.swift:400]: Does Vimeo, Twitter and Reddit all do this??? 😱

Comment (kylehickinson) [reply Shared/Extensions/URLExtensions.swift:400]: Also shouldn't we only be doing this for Mobile YouTube/Twitch (m.youtube.com)

File: Shared/Extensions/URLExtensions.swift
Comment (kylehickinson) [inline Shared/Extensions/URLExtensions.swift:395]: Should probably say this is rewards specific somehow. Since there are obviously tons of other media URLs.

File: Client/Frontend/Browser/BrowserViewController.swift
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift:1111]: Can we add a comment about what this is for and also add a line of whitespace below it

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: Shared/Extensions/URLExtensions.swift
Comment (danishjafri88) [reply Shared/Extensions/URLExtensions.swift:400]: @kylehickinson yeah all of them have the same issue. Of the rewards media list only Github did not do this.

Comment (danishjafri88) [reply Shared/Extensions/URLExtensions.swift:400]: @kylehickinson for desktop too they behave same.

Comment (kylehickinson) [reply Shared/Extensions/URLExtensions.swift:400]: Hmm, I wonder if Chromium sees these changes and handles them properly and its just WKWebView that doesn't

Comment (danishjafri88) [reply Shared/Extensions/URLExtensions.swift:400]: I think wkwebview should be aware of this too, they are just not providing a callback for the same. Apart from url change kvo

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**
